R-package/src/Makevars.in

```diff
@@ -0,0 +1,56 @@
+CXX_STD = CXX11
+
+PKGROOT=.
+
+LGB_CPPFLAGS = \
+    @LGB_CPPFLAGS@ \
+    -DUSE_SOCKET \
+    -DLGB_R_BUILD
+
+CPICFLAGS = -fPIC
+
+PKG_CPPFLAGS = \
+    -I$(PKGROOT)/include \
+    $(LGB_CPPFLAGS)
+
+PKG_CXXFLAGS = \
+    @OPENMP_CXXFLAGS@ \
+    -pthread
+
+PKG_LIBS = \
+    @OPENMP_CXXFLAGS@ \
+    @OPENMP_LIB@ \
+    -pthread
+
+OBJECTS = \
```

Comment:
I think we can benefit from using wildcard here: https://stackoverflow.com/a/49557364.

Reply:
oh cool!

I see other packages on CRAN using `wildcard`, so I think it's ok: https://github.com/search?l=Makefile&p=1&q=org%3Acran+wildcard&type=Code

I just tried this in https://github.com/microsoft/LightGBM/pull/3188/commits/53108bd3503f16ad1b7a2003f29ad017944ab108

I couldn't figure out how to match all source files recursively, but this is still an improvement on listing every individual object:

```makefile
SOURCES = $(wildcard *.cpp) \
          $(wildcard application/*.cpp) \
          $(wildcard boosting/*.cpp) \
          $(wildcard include/*.cpp) \
          $(wildcard io/*.cpp) \
          $(wildcard metric/*.cpp) \
          $(wildcard network/*.cpp) \
          $(wildcard objective/*.cpp) \
          $(wildcard treelearner/*.cpp)

OBJECTS = $(SOURCES:.cpp=.o)
```

Reply:
> I couldn't figure out how to match all source files recursively

Try this please: https://stackoverflow.com/a/4038459

Reply:
ooooooh thanks! That exact answer seems like it might not be portable, but I was able to piece together things from different answers there and this seems to be working (at least on my Mac):

```makefile
SOURCES = $(wildcard *.cpp) $(wildcard */*.cpp)
```

Trying now on CI

Reply:
it looks like we cannot use this pattern :/

![image](https://user-images.githubusercontent.com/7608904/86705542-982c3b00-bfdb-11ea-95ad-1de79bff9bf8.png)

Here is the relevant section from "Writing R Extensions", ["Writing Portable Packages"](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Writing-portable-packages)

![image](https://user-images.githubusercontent.com/7608904/86705894-f0633d00-bfdb-11ea-8723-cc209f431345.png)

I think we should list the objects manually for now. I looked around github.com/cran and it seems like many packages do that.

Reply:
Ah, does it mean that we are not allowed even to use [non-recursive `wildcard`](https://github.com/microsoft/LightGBM/pull/3188#discussion_r449733567)?

Reply:
I just saw a post on the `r-devel` mailing list this morning about this exact thing, and the advice was the same as above...list moderators told the poster they should just explicitly list the objects.

I think that as long as we have high test coverage in the R package, we'll always catch the case in CI where a new source file has been introduced but isn't being compiled.

I did not notice this line in `Writing R Extensions` until re-reading what I posted above though!

> Since the only viable make for Windows is is GNU make, it is permissible to use GNU extensions in files Makevars.win or Makefile.win

I think that means we could use the wildcard approach at least for Windows! I'll try that.

Reply:
> list moderators told the poster they should just explicitly list the objects.

It's a pity!

> I think that means we could use the wildcard approach at least for Windows!

Hm, I'm not sure it does woth it ("all or nothing")...

Reply:
yeah I think I agree with you, actually. I'd rather that both of them look the same.